[ZEPPELIN-2214] Set npm installer default npm registry

necosta opened the pull request:
### What is this PR for?
Complements ZEPPELIN-2278, this is a fix for "connection timeout" bug when trying to run "npm install" behind a proxy. Instead of connecting to the official nodejs site from behind a proxy, we should use the already soft-coded env variable ZEPPELIN_HELIUM_NPM_REGISTRY repository.

### What type of PR is it?
Bug Fix

### ToDos
* [x] - Soft-code npm folder name as well (currently hardcoded "npm") - To be agreed on...

### What is the Jira issue?
* [ZEPPELIN-2214 - Bug on connecting to the official nodejs site from behind a proxy]

### How should this be tested?
Enable any Helium plugin from behind a proxy

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? No

andreaTP commented:
Since npm registry name can be changed arbitrarily it's worth to add a configuration key for the hard-coded `"/npm/-/"`

necosta commented:
Travis is happy, Jenkins is unhappy. Anyone knows why?
```
python ./travis_check.py
Traceback (most recent call last):
  File "./travis_check.py", line 35, in <module>
    author = sys.argv[1]
IndexError: list index out of range]
```

1ambda commented:
@necosta 

What is your fork name? It should be `zeppelin`.

1ambda commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/helium/HeliumBundleFactory.java:
```diff
@@ -101,8 +101,8 @@ public HeliumBundleFactory(
     this.yarnCacheDir = new File(heliumLocalRepoDirectory, YARN_CACHE_DIR);
-    this.conf = conf;
-    this.defaultNpmRegistryUrl = conf.getHeliumNpmRegistry();
+    this.defaultNpmRegistryUrl = conf.getHeliumNpmRegistryUrl();
```
We need proxies for 

- node installation 
- npm installation (not npm packages, npm itself)
- yarn installation

https://github.com/eirslett/frontend-maven-plugin#proxy-settings

necosta commented:
Thanks @andreaTP , @1ambda , @felixcheung 

@andreaTP , I confused everyone with soft-coding the npm value. "npm" is the npm package name. So it should remain hard-coded.

@1ambda , thanks for your help. It's my first pull request so might have some issues with the CI. Answering your question, I believe I forked from zeppelin as I'm making the pull request from nokia (remote) / zeppelin (repo) / zeppelin2214 (branch). Let me know if I'm still missing something.

@felixcheung , @1ambda , ignore Npm registry repo. Latest version has now soft-coded Node registry, NPM registry and YARN registry. Node and Yarn are repositories, not registries, but with a repository manager, they become registry paths.

1ambda commented:
@necosta Thanks for quick updating! 

I have one more question.  **Is there a proxy option for npm packages? (not npm itself)**
Since if we an enable helium package, it will download specified (required) npm packages.

necosta commented:
@1ambda , thanks for the quick feedback :)
This ticket has related ticket ZEPPELIN-2152 which has a recent comment related to your question:
https://github.com/NohSeho/zeppelin/commit/032e120d06436e8b6c6b3b881efd20ba2c828fb2
Not sure what is the status of it (if under review or not proposed as a pull request yet) but I think we can separate issues.
This is, this ticket to configure the call to a repository manager, ticket ZEPPELIN-2152 to handle proxy configuration. What do you think?

1ambda commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java:
```diff
@@ -643,7 +651,10 @@ public String getJettyName() {
     ZEPPELIN_HELIUM_REGISTRY("zeppelin.helium.registry", "helium," + HELIUM_PACKAGE_DEFAULT_URL),
+    ZEPPELIN_HELIUM_NODE_REGISTRY("zeppelin.helium.node.registry", "https://nodejs.org/dist/"),
     ZEPPELIN_HELIUM_NPM_REGISTRY("zeppelin.helium.npm.registry", "http://registry.npmjs.org/"),
+    ZEPPELIN_HELIUM_YARN_REGISTRY("zeppelin.helium.yarn.registry",
```
Can we change the variable name from `registry` to `installer.url`? 

- https://github.com/eirslett/frontend-maven-plugin/blob/master/frontend-maven-plugin/src/main/java/com/github/eirslett/maven/plugins/frontend/mojo/InstallNodeAndNpmMojo.java#L5

1ambda commented:
Yes, we can split. Then, 

- Let's make CI green
- I left one comment.

necosta commented:
@1ambda, latest updates:
 - Renamed "registry" to "installer.url" as suggested in your comment
 - Fixed CI
 - "ZEPPELIN-2214 is about setting proxy for Helium. But Actually, this PR it not related with proxy itself. Just for configurable URLs." - True in a way... Should I create a new ticket? Or should we update ZEPPELIN-2214 description? What do you suggest? Thanks.

felixcheung commented on conf/zeppelin-env.cmd.template:
```diff
@@ -39,7 +39,9 @@ REM set ZEPPELIN_IDENT_STRING   		REM A string representing this instance of zep
 REM set ZEPPELIN_NICENESS       		REM The scheduling priority for daemons. Defaults to 0.
 REM set ZEPPELIN_INTERPRETER_LOCALREPO         REM Local repository for interpreter's additional dependency loading
 REM set ZEPPELIN_INTERPRETER_DEP_MVNREPO       REM Maven principal repository for interpreter's additional dependency loading
+REM set ZEPPELIN_HELIUM_NODE_INSTALLER_URL          REM Remote Node installer url for Helium dependency loader
```
align the space?

necosta replied:
Done

felixcheung commented on docs/install/configuration.md:
```diff
@@ -282,12 +282,24 @@ If both are defined, then the **environment variables** will take priority.
+      <td><h6 class="properties">ZEPPELIN_HELIUM_YARN_INSTALLER_URL</h6></td>
+      <td><h6 class="properties">zeppelin.helium.yarn.installer url</h6></td>
+      <td>https://github.com/yarnpkg/yarn/releases/download/</td>
+      <td>Remote Yarn installer url for Helium dependency loader</td>
```
maybe `yarn` is a bit confusing? I guess with a lot of the projects going on here I mostly associate `yarn` with `hadoop yarn`

not sure if there is a better way to describe this though

1ambda replied:
then, we can use `yarnpkg` instead of. (https://yarnpkg.com/lang/en/)

necosta replied:
Done

1ambda commented:
Personally, I hope we have a proxy option instead of configurations for those URLs.
That's because, it's hard to configure every URLs for users. (not scalable solution) There might be more URLs except for helium related thins.  (e.g maven dep loader, `%spark.dep`, ...)

If we have proxy support in Zeppelin, we just need to set that configuration only. 

> Does it make sense to use the well known environment variables http_proxy and no_proxy? They are usually configured correctly already on all the machines that require proxy settings. If there is a use case to have different proxy settings for the node installer and the npm installer (although I don't have one personally right now), we can still provide separate configuration properties. But it would certainly come in handy to use http_proxy for all of them if it is set and not overwritten.

1ambda commented on conf/zeppelin-env.cmd.template:
```diff
@@ -39,7 +39,9 @@ REM set ZEPPELIN_IDENT_STRING   		REM A string representing this instance of zep
 REM set ZEPPELIN_INTERPRETER_LOCALREPO         REM Local repository for interpreter's additional dependency loading
 REM set ZEPPELIN_INTERPRETER_DEP_MVNREPO       REM Maven principal repository for interpreter's additional dependency loading
+REM set ZEPPELIN_HELIUM_NODE_INSTALLER_URL     REM Remote Node installer url for Helium dependency loader
 REM set ZEPPELIN_HELIUM_NPM_REGISTRY           REM Remote Npm registry for Helium dependency loader
```
Should be `ZEPPELIN_HELIUM_NPM_INSTALLER_URL`?

necosta replied:
I avoided changing this one as it will "break" releases already using ZEPPELIN_HELIUM_NPM_REGISTRY (config needs to be manually adjusted). 
But happy to rename it. Is there anything "special" I need to do to inform community that they need to rename ZEPPELIN_HELIUM_NPM_REGISTRY to ZEPPELIN_HELIUM_NPM_INSTALLER_URL? Thanks.

1ambda replied:
Helium is still experimental. IMO, it's ok to change. 
We can write down the change in https://github.com/apache/zeppelin/blob/master/docs/install/upgrade.md

1ambda commented on zeppelin-zengine/src/main/java/org/apache/zeppelin/conf/ZeppelinConfiguration.java:
```diff
@@ -424,8 +424,16 @@ public String getHeliumRegistry() {
+  }
+
+  public String getHeliumNpmInstallerUrl() {
+    return getString(ConfVars.ZEPPELIN_HELIUM_REGISTRY);
```
Here too. `ZEPPELIN_HELIUM_NPM_INSTALLER_URL`

1ambda commented:
If we have both, then it's better.

necosta commented:
@1ambda ,
Renamed "zeppelin.helium.npm.registry" to "zeppelin.helium.npm.installer.url"
I understand there is no need to update file "upgrade.md" as the change in conf value is inside the same version (0.8)
Ready for review.  Thanks

necosta commented:
Hi @1ambda and @Leemoonsoo ,
Can you have another look at this one?
After rebasing latest master (now contains [ZEPPELIN-2152](https://issues.apache.org/jira/browse/ZEPPELIN-2152)) and squashing my commits, this is the version I would like to get approved.
There is a lot of info. in the above comments. Any question, let me know and thank you.
By the way, I linked [ZEPPELIN-2227](https://issues.apache.org/jira/browse/ZEPPELIN-2227) as duplicate to this one ([ZEPPELIN-2214](https://issues.apache.org/jira/browse/ZEPPELIN-2214)).

necosta commented:
ping

Leemoonsoo commented:
Tested and LGTM. Thanks @necosta for the improvement!
Merge to master if no further comment.
